code/game/machinery/telecomms/machine_interactions.dm

```diff
@@ -27,74 +29,176 @@
+				id = params["value"]
+				. = TRUE
+		if("network")
+			if(params["value"])
+				if(length(params["value"]) > 15)
+					return
+				else
+					for(var/obj/machinery/telecomms/T in links)
+						T.links.Remove(src)
+					network = params["value"]
+					links = list()
+					log_game("[key_name(operator)] has changed the network for [src] at [AREACOORD(src)] to [network].")
+					. = TRUE
+		if("tempfreq")
+			if(params["value"])
+				tempfreq = text2num(params["value"])
+		if("freq")
+			var/newfreq = tempfreq * 10
+			if(newfreq == FREQ_SYNDICATE)
+				to_chat(operator, "<span class='warning'>Error: Interference preventing filtering frequency: \"[newfreq / 10] GHz\"</span>")
+				playsound(src, 'sound/machines/buzz-sigh.ogg', 50, TRUE)
 			else
-				dat += "<br><br>MULTITOOL BUFFER: <a href='?src=[REF(src)];buffer=1'>\[Add Machine\]</a>"
+				if(!(newfreq in freq_listening) && newfreq < 10000)
+					freq_listening.Add(newfreq)
+					log_game("[key_name(operator)] added frequency [newfreq] for [src] at [AREACOORD(src)].")
+					. = TRUE
+		if("delete")
+			freq_listening.Remove(params["value"])
+			log_game("[key_name(operator)] added removed frequency [params["value"]] for [src] at [AREACOORD(src)].")
+			. = TRUE
+		if("unlink")
+			var/obj/machinery/telecomms/T = links[text2num(params["value"])]
+			if(T)
+				// Remove link entries from both T and src.
+				if(T.links)
+					T.links.Remove(src)
+				links.Remove(T)
+				log_game("[key_name(operator)] unlinked [src] and [T] at [AREACOORD(src)].")
+				. = TRUE
+		if("link")
+			if(heldmultitool)
+				var/obj/machinery/telecomms/T = heldmultitool.buffer
+				if(istype(T) && T != src)
+					if(!(src in T.links))
+						T.links += src
+					if(!(T in links))
+						links += T
+						log_game("[key_name(operator)] linked [src] for [T] at [AREACOORD(src)].")
+						. = TRUE
+		if("buffer")
+			heldmultitool.buffer = src
+			. = TRUE
+		if("flush")
+			heldmultitool.buffer = null
+			. = TRUE
+
+	add_act(action, params)
+	. = TRUE
+
+/obj/machinery/telecomms/proc/add_option()
+	return
 
-	dat += "</font>"
-	temp = ""
-	user << browse(dat, "window=tcommachine;size=520x500;can_resize=0")
-	onclose(user, "tcommachine")
-	return TRUE
+/obj/machinery/telecomms/bus/add_option()
+	var/list/data = list()
+	data["type"] = "bus"
+	data["changefrequency"] = change_frequency
+	return data
+
+/obj/machinery/telecomms/relay/add_option()
+	var/list/data = list()
+	data["type"] = "relay"
+	data["broadcasting"] = broadcasting
+	data["receiving"] = receiving
+	return data
+
+/obj/machinery/telecomms/proc/add_act(action, params)
+
+/obj/machinery/telecomms/relay/add_act(action, params)
+	switch(action)
+		if("broadcast")
+			broadcasting = !broadcasting
+			. = TRUE
+		if("receive")
+			receiving = !receiving
+			. = TRUE
+
+/obj/machinery/telecomms/bus/add_act(action, params)
+	switch(action)
+		if("change_freq")
+			var/newfreq = params["value"]
+			if(newfreq)
+				if(findtext(num2text(newfreq), "."))
```

Comment:
What happens if you pass an integer frequency with no dot after a division by 10? I'd say just stick to one format of frequency, and where necessary, divide or multiply by 10.